File: movieIos/MovieCatalogue/App/Features/Home/ViewModel/HomeViewModel.swift
//
//  HomeViewModel.swift
//  MovieCatalogue
//
//  Created by Uwais Alqadri on 26/09/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import RxSwift
import RxRelay
import KotlinCore
import KMPNativeCoroutinesRxSwift

class HomeViewModel {

  let popularMovies: BehaviorRelay<[Movie]> = .init(value: [Movie]())
  let upComingMovies: BehaviorRelay<[Movie]> = .init(value: [Movie]())
  let bannerMovies: BehaviorRelay<[Movie]> = .init(value: [Movie]())

  private let disposeBag = DisposeBag()
  private let popularUseCase: GetPopularMovieUseCase
  private let upComingUseCase: GetUpComingMovieUseCase
  private let bannerUseCase: GetBannerMovieUseCase

  init(popularUseCase: GetPopularMovieUseCase, upComingUseCase: GetUpComingMovieUseCase, bannerUseCase: GetBannerMovieUseCase) {
    self.popularUseCase = popularUseCase
    self.upComingUseCase = upComingUseCase
    self.bannerUseCase = bannerUseCase
  }

  func requestPopularMovie() {
    createObservable(for: popularUseCase.executeNative(year: Extensions().currentYear, page: 1, sortBy: .popularity))
      .observe(on: MainScheduler.instance)
      .subscribe(onNext: { [weak self] result in
        self?.popularMovies.accept(Array(result[0...9]))
      }).disposed(by: disposeBag)
  }

  func requestUpComingMovie() {
    createObservable(for: upComingUseCase.executeNative(year: Extensions().nextYear, page: 1, sortBy: .popularity))
      .observe(on: MainScheduler.instance)
      .subscribe(onNext: { [weak self] result in
        self?.upComingMovies.accept(Array(result[0...9]))
      }).disposed(by: disposeBag)
  }

  func requestBannerMovie() {
    createObservable(for: bannerUseCase.executeNative(year: Extensions().currentYear, page: 1, sortBy: .release_))
      .observe(on: MainScheduler.instance)
      .subscribe(onNext: { [weak self] result in
        self?.bannerMovies.accept(Array(result[0...2]))
      }).disposed(by: disposeBag)
  }
}

File: movieAndroid/src/main/java/com/uwaisalqadri/moviecatalogue/android/ui/home/HomeViewModel.kt
package com.uwaisalqadri.moviecatalogue.android.ui.home

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.uwaisalqadri.moviecatalogue.domain.model.Movie
import com.uwaisalqadri.moviecatalogue.domain.usecase.home.GetBannerMovieUseCase
import com.uwaisalqadri.moviecatalogue.domain.usecase.home.GetPopularMovieUseCase
import com.uwaisalqadri.moviecatalogue.domain.usecase.home.GetUpComingMovieUseCase
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

/**
 * Created by Uwais Alqadri on September 24, 2021
 */
class HomeViewModel(
	private val popularUseCase: GetPopularMovieUseCase,
	private val upComingUseCase: GetUpComingMovieUseCase,
	private val bannerUseCase: GetBannerMovieUseCase
): ViewModel() {

	val popularMovie = MutableLiveData<List<Movie>>()
	val bannerMovie = MutableLiveData<List<Movie>>()
	val upComingMovie = MutableLiveData<List<Movie>>()

	val loading = MutableLiveData(true)

	init {
		requestBannerMovie()
		requestPopularMovie()
		requestUpComingMovie()
	}

	private fun requestPopularMovie() = viewModelScope.launch {
		popularUseCase.execute().collect {
			loading.value = false
			popularMovie.value = it.slice(0 until 10)
		}
	}

	private fun requestUpComingMovie() = viewModelScope.launch {
		upComingUseCase.execute().collect {
			loading.value = false
			upComingMovie.value = it.slice(0 until 10)
		}
	}

	private fun requestBannerMovie() = viewModelScope.launch {
		bannerUseCase.execute().collect {
			loading.value = false
			bannerMovie.value = it.slice(0 until 3)
		}
	}
}